app/code/Magento/InventoryShipping/Plugin/Sales/Shipment/AssignSourceCodeToShipmentPlugin.php

```diff
@@ -0,0 +1,116 @@
+use Magento\InventoryApi\Api\GetStockSourceLinksInterface;
+use Magento\InventoryApi\Api\Data\StockSourceLinkInterface;
+
+class AssignSourceCodeToShipmentPlugin
+{
+    /**
+     * @var RequestInterface
+     */
+    private $request;
+
+    /**
+     * @var ShipmentExtensionFactory
+     */
+    private $shipmentExtensionFactory;
+
+    /**
+     * @var StockByWebsiteIdResolver
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var GetStockSourceLinksInterface
+     */
+    private $getStockSourceLinks;
+
+    /**
+     * AssignSourceCodeToShipmentPlugin constructor.
+     * @param RequestInterface $request
+     * @param ShipmentExtensionFactory $shipmentExtensionFactory
+     * @param StockByWebsiteIdResolver $stockByWebsiteIdResolver
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param GetStockSourceLinksInterface $getStockSourceLinks
+     */
+    public function __construct(
+        RequestInterface $request,
+        ShipmentExtensionFactory $shipmentExtensionFactory,
+        StockByWebsiteIdResolver $stockByWebsiteIdResolver,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        GetStockSourceLinksInterface $getStockSourceLinks
+    ) {
+        $this->request = $request;
+        $this->shipmentExtensionFactory = $shipmentExtensionFactory;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->getStockSourceLinks = $getStockSourceLinks;
+    }
+
+    /**
+     * @param ShipmentFactory $subject
+     * @param ShipmentInterface $shipment
+     * @param Order $order
+     * @return ShipmentInterface
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterCreate(ShipmentFactory $subject, ShipmentInterface $shipment, Order $order)
+    {
+        $sourceCode = $this->request->getParam('source_code');
+        if (empty($sourceCode)) {
+            $websiteId = $order->getStore()->getWebsiteId();
+            $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
+            $sources = $this->getAssignedSourcesForStock($stockId);
+            if (!empty($sources) && count($sources) == 1) {
+                $sourceCode = $sources[0];
+            }
+        }
+        $shipmentExtension = $shipment->getExtensionAttributes();
+
+        if (empty($shipmentExtension)) {
+            $shipmentExtension = $this->shipmentExtensionFactory->create();
+        }
+        $shipmentExtension->setSourceCode($sourceCode);
+        $shipment->setExtensionAttributes($shipmentExtension);
+
+        return $shipment;
+    }
+
+    /**
+     * Retrieves sources that are assigned to $stockId
+     *
+     * @param int $stockId
+     * @return StockSourceLinkInterface[]
+     */
+    private function getAssignedSourcesForStock(int $stockId): array
```

Comment:
you can re-use this service instead
`\Magento\InventoryApi\Api\GetSourcesAssignedToStockOrderedByPriorityInterface`